dask/order.py

```diff
@@ -93,28 +94,28 @@ def ndependents(dependencies, dependents):
     >>> sorted(ndependents(dependencies, dependents).items())
     [('a', 3), ('b', 2), ('c', 1)]
     """
-    result = dict()
-
-    roots = [k for k, v in dependents.items() if not v]
-
-    result.update(dict((r, 1) for r in roots))
-
+    result = dict((k, 1) for k, v in dependents.items() if not v)
     leaves = [k for k, v in dependencies.items() if not v]
-
     for leaf in leaves:
-        _ndependents(leaf, result, dependencies, dependents)
-
+        _ndependents(leaf, result, dependents)
     return result
 
 
-def _ndependents(key, result, dependencies, dependents):
+def _ndependents(key, result, dependents):
     """ Helper function for ndependents """
-    if key not in result:
+    # 2 is the default value for non root because if we're 2 then we won't be in result
+    # and we won't get value when during the call the to setdefault
+    stack = [(key, 2)]
+    while stack:
+        key, value = stack.pop()
         deps = dependents[key]
-        result[key] = sum(
-            [_ndependents(k, result, dependencies, dependents)
-             for k in deps]) + 1
-    return result[key]
+        ndeps = 0
+        for dep in deps:
+            val = result.setdefault(dep, value)
+            stack.append((dep, val))
```

Comment:
This still seems to add all dependents many times onto the stack.  I would expect there to be some check like 

```
if we haven't seen this dependency before:
    add it on to the stack
```

I may be wrong about this though.